vql/linux/audit/audit_service.go

```diff
@@ -0,0 +1,1018 @@
+		self.Debug("audit: shutdown watcher starting")
+		defer self.Debug("audit: shutdown watcher exited")
+
+		select {
+			// If we exit the main event loop normally
+			case <-ctx.Done():
+				break
+			// If any of the goroutines exits abnormally
+			case <-grpctx.Done():
+				break
+		}
+
+		err := grp.Wait()
+		if !errors.Is(err, context.Canceled) {
+			self.logger.Info("audit: shutting down due to error ; err=%s", err)
+		}
+
+		self.shutdown()
+	}()
+
+	self.cancelService = cancel
+	return nil
+}
+
+func (self *auditService) shutdown() {
+	// If we're shutting down due to error, we'll still have subscribed callers
+	self.subscriberLock.Lock()
+	for _, subscriber := range self.subscribers {
+		self.unsubscribe(subscriber, true)
+	}
+	self.subscribers = []*subscriber{}
+	self.subscriberLock.Unlock()
+
+	self.reassembler.Close()
+	self.reassembler = nil
+
+	self.commandClient.Close()
+	self.commandClient = nil
+
+	self.messageQueue.Close()
+	self.messageQueue = nil
+
+	close(self.logChannel)
+	close(self.checkerChannel)
+
+	self.bannedRules = map[string]*AuditRule{}
+	self.rules = map[string]*RefcountedAuditRule{}
+
+	self.logger.Info("audit: Shut down audit service")
+
+	self.serviceLock.Lock()
+	defer self.serviceLock.Unlock()
+
+	self.running = false
+	self.shuttingDown = false
+	self.serviceWg.Done()
+}
+
+func openAuditListenerSocket() (int, error) {
+	sockFd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, syscall.NETLINK_AUDIT)
+	if err != nil {
+		return -1, err
+	}
+
+	src := &syscall.SockaddrNetlink{Family: syscall.AF_NETLINK,
+					Groups: unix.AUDIT_NLGRP_READLOG}
+
+	err = syscall.Bind(sockFd, src)
+	if err != nil {
+		syscall.Close(sockFd)
+		return -1, fmt.Errorf("Could not bind to netlink socket: %w", err)
+	}
+
+	return sockFd, nil
+}
+
+func openEpollDescriptor(fd int) (int, error) {
+	pollFd, err := unix.EpollCreate1(0)
+	if err != nil {
+		return -1, err
+	}
+
+	err = unix.EpollCtl(pollFd, unix.EPOLL_CTL_ADD, fd,
+			    &unix.EpollEvent{Events: unix.POLLIN | unix.POLLHUP,
+			    Fd: int32(fd)})
+	if err != nil {
+		syscall.Close(pollFd)
+		return -1, err
+	}
+
+	return pollFd, nil
+}
+
+type RawAuditMessageBuf struct {
+	Message		libaudit.RawAuditMessage
+	Data		[]byte
+}
+
+func (self *auditService) acceptEvents(ctx context.Context, sockFd int) error {
+	receivedCount := 0
+	discardedCount := 0
+	queuedCount := 0
+
+	// We're in non-blocking mode.  Try to get all of the events we can in one go.
+	var err error
+	for {
+		err = ctx.Err()
+		if err != nil {
+			break
+		}
+
+		buf := self.rawBufPool.Get().([]byte)
+		msgType, size, err := self.receiveMessageBuf(sockFd, buf)
+		if err != nil {
+			self.rawBufPool.Put(buf)
+			break
+		}
+
+		receivedCount += 1
+
+		// Messages from 1300-2999 are valid audit messages.
+		if msgType < auparse.AUDIT_USER_AUTH || msgType > auparse.AUDIT_LAST_USER_MSG2 {
+			self.rawBufPool.Put(buf)
+			discardedCount += 1
+			continue
+		}
+
+		self.messageQueue.Send(buf[:size])
```

Comment:
I think this needs to be `self.messageQueue.Send(buf)` so Put() in queuemainEventLoop() will put the whole buffer back in the pool. 
https://gist.github.com/jaysoncena/f2c98cba63f26a181c0ebb216b80d3de#file-bytepool1-go-L20-L24

Reply:
I spent some more time looking at this. Sending to messageQueue means it might goes to disk, and I think we read back from disk into some other address, and that should not be put back into the pool.

I found this fixes the problem
```
               bufCopy := make([]byte, size)
               copy(bufCopy, buf)
               self.rawBufPool.Put(buf)
               self.messageQueue.Send(bufCopy)
```
and remove `self.rawBufPool.Put(buf)` from `mainEventLoop()`. 

But maybe this defeats the benefit of using sync.Pool in the first place.

Reply:
I was thinking about this as well and I think it's clear it's not fully baked. I think the sync pool still makes sense, but the lifecycle needs to be fixed. I've mostly finished up reworking the bpf stuff to be (hopefully) acceptable for submission and will take another look at this.

Reply:
This one has been resolved.